File: communication/mqtt/src/main/java/org/openbase/jul/communication/mqtt/RPCServer.java
package org.openbase.jul.communication.mqtt;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import com.hivemq.client.mqtt.datatypes.MqttQos;
import com.hivemq.client.mqtt.mqtt3.Mqtt3AsyncClient;
import com.hivemq.client.mqtt.mqtt5.Mqtt5AsyncClient;
import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;

import java.nio.charset.StandardCharsets;
import java.util.function.Consumer;

public class RPCServer {

    private final Mqtt5AsyncClient mqttClient;
    private final String topic;

    public RPCServer(final Mqtt5AsyncClient client, final String topic) {
        this.mqttClient = client;
        this.topic = topic + "/rpc";
    }

    public void activate() {
        this.mqttClient.subscribeWith()
                .topicFilter(topic)
                .qos(MqttQos.EXACTLY_ONCE)
                .callback(new InternalSubscriber(this.mqttClient, this.topic))
                .send();
    }

    public void deactivate() {
        this.mqttClient.unsubscribeWith()
                .topicFilter(topic)
                .send();
    }
}

class InternalSubscriber implements Consumer<Mqtt5Publish> {

    private final Mqtt5AsyncClient mqttClient;
    private String topic;

    public InternalSubscriber(final Mqtt5AsyncClient mqttClient, final String topic) {
        this.mqttClient = mqttClient;
        this.topic = topic;
    }

    @Override
    public void accept(Mqtt5Publish mqtt5Publish) {
        ResponseType.Response.Builder responseBuilder = ResponseType.Response.newBuilder();
        try {
            RequestType.Request request = RequestType.Request.parseFrom(mqtt5Publish.getPayloadAsBytes());

            this.topic += "/" + request.getId();

            responseBuilder.setStatus(ResponseType.Response.Status.ACKNOWLEDGED);
            responseBuilder.setId(request.getId());
            mqttClient.publishWith()
                    .topic(this.topic)
                    .qos(MqttQos.EXACTLY_ONCE)
                    .payload(responseBuilder.build().toByteArray())
                    .send(); //TODO call get?

            //TODO open thread to send PROGRESSING message periodically

            try {
                //TODO call method as defined in request.getMethodName();
                request.getMethodName();
                responseBuilder.setStatus(ResponseType.Response.Status.FINISHED);
                responseBuilder.setResult(ByteString.copyFrom(request.getMethodName().getBytes(StandardCharsets.UTF_8)));
            } catch (Exception ex) {
                responseBuilder.setStatus(ResponseType.Response.Status.FINISHED);
                responseBuilder.setError(ex.getMessage()); //TODO: build message accordingly from stackstrace...
            }

            mqttClient.publishWith()
                    .topic(this.topic)
                    .qos(MqttQos.EXACTLY_ONCE)
                    .payload(responseBuilder.build().toByteArray())
                    .send(); //TODO call get?
        } catch (InvalidProtocolBufferException e) {
            // We cannot inform the client here because we do not know
            // on which topic the client listens, so just log the error
            //TODO: log err
        }
    }
}

interface Callback<R extends Message> {

    R call();
}

File: communication/mqtt/src/main/java/org/openbase/jul/communication/mqtt/RPCServer.kt
package org.openbase.jul.communication.mqtt

import com.hivemq.client.mqtt.datatypes.MqttQos
import com.hivemq.client.mqtt.mqtt5.Mqtt5AsyncClient
import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish
import com.hivemq.client.mqtt.mqtt5.message.subscribe.Mqtt5Subscribe
import com.hivemq.client.mqtt.mqtt5.message.unsubscribe.Mqtt5Unsubscribe
import org.openbase.jul.exception.CouldNotPerformException
import org.openbase.jul.schedule.GlobalCachedExecutorService
import org.openbase.type.communication.mqtt.RequestType
import org.openbase.type.communication.mqtt.ResponseType
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*
import java.util.concurrent.Future
import java.util.function.BiConsumer

class RPCServer(private val mqttClient: Mqtt5AsyncClient, topic: String) {

    private val topic: String = "$topic/rpc"
    private val methods: HashMap<String, RPCMethod> = HashMap()

    private var activationFuture: Future<out Any>? = null
    val isActive: Boolean = (activationFuture != null && activationFuture!!.isDone && !activationFuture!!.isCancelled)

    fun activate(): Future<out Any> {
        if (activationFuture != null && !activationFuture!!.isDone) {
            return activationFuture!!
        }

        activationFuture = mqttClient.subscribe(
            Mqtt5Subscribe.builder()
                .topicFilter(topic)
                .qos(MqttQos.EXACTLY_ONCE)
                .build(),
            { mqtt5Publish -> handleRemoteCall(mqtt5Publish) },
            GlobalCachedExecutorService.getInstance().executorService
        )
        return activationFuture!!
    }

    fun deactivate(): Future<out Any> {
        return mqttClient.unsubscribe(
            Mqtt5Unsubscribe.builder()
                .topicFilter(topic)
                .build()
        )
    }

    fun addMethod(method: Method, instance: Any) {
        methods[method.name] = RPCMethod(method, instance);
    }

    private fun handleRemoteCall(mqtt5Publish: Mqtt5Publish) {
        val responseBuilder = ResponseType.Response.newBuilder()
        val request = RequestType.Request.parseFrom(mqtt5Publish.payloadAsBytes)

        // make sure that the request id is a valid uuid so that request
        // collisions are unlikely
        responseBuilder.id = UUID.fromString(request.id).toString()
        val requestTopic = topic + "/" + responseBuilder.id
        val mqttResponseBuilder = Mqtt5Publish.builder()
            .topic(requestTopic)
            .qos(MqttQos.EXACTLY_ONCE)

        responseBuilder.status = ResponseType.Response.Status.ACKNOWLEDGED
        responseBuilder.id = request.id
        mqttClient.publish(
            mqttResponseBuilder
                .payload(responseBuilder.build().toByteArray())
                .build()
        ) //TODO call get?

        if (request.methodName !in methods) {
            responseBuilder.status = ResponseType.Response.Status.FINISHED;
            responseBuilder.error = "Method ${request.methodName} is not available";

            mqttClient.publish(
                mqttResponseBuilder
                    .payload(responseBuilder.build().toByteArray())
                    .build()
            ) //TODO call get?
            return;
        }

        val method = methods[request.methodName]!!;

        //TODO open thread to send PROGRESSING message periodically
        responseBuilder.status = ResponseType.Response.Status.FINISHED
        try {
            val result = method.invoke(request.paramsList);
            responseBuilder.result = result
        } catch (ex: InvocationTargetException) {
            responseBuilder.error = ex.cause!!.message //TODO: build message accordingly from stackstrace...
        } catch (ex: CouldNotPerformException) {
            responseBuilder.error = ex.message
        } catch (ex: Exception) {
            responseBuilder.error = "Server error ${ex.message}"
        }

        mqttClient.publish(
            mqttResponseBuilder
                .payload(responseBuilder.build().toByteArray())
                .build()
        ) //TODO call get?
    }
}